### ISSUE TITLE
Question - how to get app URL from CLI?

### ISSUE THREAD

**Issue by garethellis36** [beporter/CakePHP-EnvAwareness#6]
Hi Brian,

I have another question for you, this time about how you get the correct app URL when in the CLI. Our app sends emails from shell scripts , and these emails contain links to pages in the app.

Our current environment detection system is pretty clunky and probably breaks a bunch of the rules you outlined in your talk! 

First of all, it checks if `$_SERVER["HTTP_HOST"]` is set. If not, it does some further `file_exists()` calls to look for files which identify which server it's on (because I hadn't heard of environment variables when I wrote this!). Based on the results of that, HTTP_HOST is then set. i.e. if `file_exists("this_is_production")` returns true then `$_SERVER["HTTP_HOST"]` is set to the production URL.

Then follows a switch block which switches on the value of `$_SERVER["HTTP_HOST"]` and sets environment vars for each possible URL.

Finally, it checks if `FULL_BASE_URL` is defined (app is still on Cake 2.2 before this was deprecated), and if not, defines it as being equal to `$_SERVER["HTTP_HOST"]`. This gives the app access to the correct app URL whether running as HTTP or CLI.

How would you handle this with your method? The only thing I can think of is to set the URL with the Configure class in app-{$env}?

**Comment by beporter** [comment]
Another good question. Yeah, this gets sticky too, but I've found the env-awareness can help here. I've had to do the same kind of thing in a nightly Shell that was generating emails and needed the base URL to be correct for the given instance of the app (dev, staging, production, etc.)

The way we approached this is to let the env you set up also define a default domain that should be associated with it. So in our `core.php` we have something like this:

``` php
/**
 * Default Site Configuration
 *
 * Any time you'd be tempted to type one of these strings directly into
 * a file, call this Configure var instead.
 */
Configure::write('Defaults', array(
    'short_name' => 'ACME',
    'long_name' => 'ACME Inc Application',
    'domain' => 'acme.com',
    'meta_description' => 'Maker of great widgets.',
    'meta_keywords' => 'acme, widgets, great',
));
```

...which of course you can override for each environment.

Then in our Shell's `initialize()` method:

``` php
    /**
     * Startup for the EmailShell
     *
     * This could also even go in your application's AppShell to let every Shell inherit it.
     */
    public function initialize() {
        parent::initialize();
        if (!defined('FULL_BASE_URL')) {
            define('FULL_BASE_URL', 'http://' . Configure::read('Defaults.domain'));
        }
        // ...
    }
```

The domain isn't used for anything else from the web browser end of things, since Cake's Router class properly uses whatever domain it's currently being served from in that context. This approach let's you customize the output for each environment, but also only engage that customization from the command line, and let's you do it cleanly. It makes the domain a **prescriptive** value that _you_ tell the app, instead of an **passive** value that the app obtains from somewhere else.

I have yet to come with a better or simpler method, but as usual if anyone can improve on this I'll happily switch myself.

**Comment by beporter** [comment]
On the `file_exists()` point too, I want to mention that this is actually a great alternative to environment variables. In fact, in some instances it's far easier. PHP running under FPM for example, does not get provided with any Apache `SetEnv` directives, so it becomes hard to provide an environment variable to the PHP runtime. Using a "marker file" is a great alternative.

As I said in my talk, it really doesn't matter what mechanism you use to detect the environment, as long as you're only having to check for it once and your app is set up to "to the right thing" even if that detection fails. Of course, the simpler and less error-prone the checking process is, the better. I like Apache SetEnv because I pretty much always have to set up a virtual host _anyway_ so it's not so much an extra step as it is a part of an existing step to include the `SetEnv APP_ENV whatever` line. 

That said, in some situations you just have to do the best you can with what you've got, but you should still strive to keep it as bulletproof as possible. For example, if I was going to use the "check a file" approach, my code would probably look something like this:

``` php
// in Config/core.php

// Replace this:
// $env = getenv('APP_ENV');

// With this:
$env = trim(@file_get_contents(dirname(__FILE__) . DS . 'APP_ENV.txt'));

// Still do the same config loading.
if (is_readable(dirname(__FILE__) . DS . "core-{$env}.php")) {
    Configure::load("core-{$env}");
}

```

Why this way?
- I'd add `Config/APP_ENV.txt` to my `.gitignore` file to keep it out of source control, so it doesn't get committed with a fixed value that might get deployed to other environments and cause problems. This forces someone to create the correct files for each other environment.
- I want to always check the same file name (because there's less chance of something going wrong, typos, etc.) and because it doesn't require if/elseif/else statements for every possible environment. Just read the same file, and get the correct env from inside. This mirrors reading the same env var, and getting the correct **value** from inside of it.
- I use the `@` suppression because I don't care if the file doesn't exist (maybe you do though) and I want to keep going assuming we're on production if reading the file fails.
- I use `trim()` to take care of stray whitespace in the file, which is more likely to happen accidentally than in an environment variable and cause problems matching environments. Which means even a file that looks like this still works:
  
  ![screenshot 2015-06-12 09 33 00](https://cloud.githubusercontent.com/assets/637270/8133558/afefb11e-10ee-11e5-86ff-cbdf1d592f8b.PNG)
- It's still just that one single line of code that makes the whole thing work! I like the elegance of that.

All in all, I think this minimizes the amount of things that can go wrong, and the amount of work I have to do manually to maintain each environment. It's a single unique word typed in the same file file for every environment.

**Comment by garethellis36** [comment]
Thanks for the detailed responses, I appreciate you taking the time. Glad to hear my file_exists() solution is not so hacky after all!

I like your proposed solution for the CLI URL. I would assume that one should set a default protocol as well in the config, e.g. 

```
/**
 * Default Site Configuration
 *
 * Any time you'd be tempted to type one of these strings directly into
 * a file, call this Configure var instead.
 */
Configure::write('Defaults', array(
    'short_name' => 'ACME',
    'long_name' => 'ACME Inc Application',
    'domain' => 'acme.com',
    'protocol' => 'http',
    'meta_description' => 'Maker of great widgets.',
    'meta_keywords' => 'acme, widgets, great',
));
```

Then in `AppShell`:

```
    /**
     * Startup for the EmailShell
     *
     * This could also even go in your application's AppShell to let every Shell inherit it.
     */
    public function initialize() {
        parent::initialize();
        if (!defined('FULL_BASE_URL')) {
            define('FULL_BASE_URL', Configure::read('Defaults.protocol') . '://' . Configure::read('Defaults.domain'));
        }
        // ...
    }
```

**Comment by beporter** [comment]
Yeah, that's an improvement in my opinion. Alternatively, you could just include the scheme in the existing var (and rename it for clarity/accuracy) if this was the only way it was ever going to be used.

``` php
Configure::write('Defaults.shellBaseUrl', 'https://acme.com');
```

Then there's a whole other discussion on the best way to "enforce" SSL in the environments that require it (like production) but gracefully leave it disabled (or optional) in dev or testing environments...
